File: Ianus/domain/LicenseValidation.ts
import { ILicense } from "./License";

// from https://developers.google.com/web/updates/2012/06/How-to-convert-ArrayBuffer-to-and-from-String
const str2ab = (str: string) => {
    const buf = new ArrayBuffer(str.length);
    const bufView = new Uint8Array(buf);
    for (let i = 0, strLen = str.length; i < strLen; i++) {
      bufView[i] = str.charCodeAt(i);
    }
    return buf;
};
 
const importRsaKey = (pem: string) => {
    // base64 decode the string to get the binary data
    const binaryDerString = window.atob(pem);
    // convert from a binary string to an ArrayBuffer
    const binaryDer = str2ab(binaryDerString);
  
    return window.crypto.subtle.importKey(
      "spki",
      binaryDer,
      {
        name: "RSASSA-PKCS1-v1_5",
        hash: "SHA-256",
      },
      true,
      ["verify"]
    );
};

const arrayBufferFromString = (str: string) => {
    const buf = new ArrayBuffer(str.length);
    const bufView = new Uint8Array(buf);
    for (let i = 0, strLen = str.length; i < strLen; i++) {
        bufView[i] = str.charCodeAt(i);
    }
    return buf;
};

const validateLicense = (license: ILicense): [boolean, string] => {
    if (!license) {
        return [false, "No license passed!"];
    }

    if (!license.sub || !license.aud || !license.iss || !license.exp) {
        return [false, "Incomplete license!"];
    }

    if (license.iss !== "XRM-OSS") {
        return [false, "Invalid license issuer"];
    }

    if (license.aud !== "XRM-OSS-AITUNE") {
        return [false, "This license is not scoped to AI Tune!"];
    }

    if (isNaN(license.exp)) {
        return [false, "License expiry is invalid!"];
    }

    if (new Date(license.exp * 1000) < new Date()) {
        return [false, "Your license has expired"];
    }

    if (license.sub !== window.location.hostname) {
        return [false, `This license is not intended for usage in ${window.location.hostname} but for ${license.sub}. You need a matching license!`];
    }

    return [true, ""];
};

// Helper for Base64url encoding, from: https://thewoods.blog/base64url/
const base64url_decode = (value: string): ArrayBuffer => {
    const m = value.length % 4;
    return Uint8Array.from(atob(
        value.replace(/-/g, '+')
            .replace(/_/g, '/')
            .padEnd(value.length + (m === 0 ? 0 : 4 - m), '=')
    ), c => c.charCodeAt(0)).buffer
};

export const checkLicense = async (publicKey: string, licenseKey: string | undefined): Promise<[string, ILicense?]> => {
    if (!licenseKey) {
        return ["No license key passed!"];
    }

    try {
        const key = await importRsaKey(publicKey);
        const parts = licenseKey.split('.');

        if (parts.length < 3) {
            return ["Incorrect license!"];
        }

        const [encodedHeaders, encodedClaims, signature] = parts;

        const plainClaims = window.atob(encodedClaims);
        const claimsJson = JSON.parse(plainClaims);

        const [contentValidationResult, contentValidationResultMessage] = validateLicense(claimsJson);

        if (!contentValidationResult) {
            return [contentValidationResultMessage];
        }

        const isLicenseSignatureValid = await window.crypto.subtle.verify(
            "RSASSA-PKCS1-v1_5",
            key,
            base64url_decode(signature),
            new TextEncoder().encode(encodedHeaders + "." + encodedClaims)
        );

        return [isLicenseSignatureValid ? "" : "License signature verification failed!", isLicenseSignatureValid ? claimsJson : undefined];
    }
    catch {
        return ["Oops, something went wrong while validating your license"];
    }
};

File: dataverse/ianus-client/LicenseValidation.cs
using System;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Security.Cryptography;
using Microsoft.Xrm.Sdk;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.OpenSsl;
using Microsoft.Xrm.Sdk.Messages;
using Microsoft.Xrm.Sdk.Query;

namespace Ianua.Ianus.Client
{
    public static class LicenseValidation
    {
        public static RSA ImportRsaPublicKey(string pem)
        {
            var rsaParams = ConvertPemToRsaParameters(pem);

            var rsa = RSA.Create();
            rsa.ImportParameters(rsaParams);

            return rsa;
        }

        private static RSAParameters ConvertPemToRsaParameters(string pem)
        {
            using (var publicKeyTextReader = new StringReader(pem))
            {
                var pemReader = new PemReader(publicKeyTextReader);
                var rsaKeyParameters = (RsaKeyParameters)pemReader.ReadObject();

                var rsaParams = new RSAParameters
                {
                    Modulus = rsaKeyParameters.Modulus.ToByteArrayUnsigned(),
                    Exponent = rsaKeyParameters.Exponent.ToByteArrayUnsigned()
                };

                return rsaParams;
            }
        }

        private static byte[] Base64UrlDecode(string input)
        {
            string base64 = input.Replace('-', '+').Replace('_', '/');

            switch (input.Length % 4)
            {
                case 2: base64 += "=="; break;
                case 3: base64 += "="; break;
            }

            return Convert.FromBase64String(base64);
        }

        private static bool VerifySignature(RSA key, byte[] data, byte[] signature)
        {
            return key.VerifyData(data, signature, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);
        }

        private static string RetrieveEnvironmentInformation (IOrganizationService service)
        {
            var request = new RetrieveMultipleRequest
            {
                Query = new QueryExpression
                {
                    EntityName = "ian_environmentinformation"
                }
            };

            var results = (RetrieveMultipleResponse) service.Execute(request);

            return results.EntityCollection.Entities.FirstOrDefault()?.GetAttributeValue<string>("ian_uniquename");
        }

        public static LicenseValidationResult ValidateLicense
        (
            string validIssuer, 
            string validAudience, 
            string dataverseOrganizationUniqueName, 
            LicenseClaims licenseClaims
        )
        {
            if (licenseClaims == null)
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = "No license passed!"
                };
            }

            if (licenseClaims.env == null || !licenseClaims.env.Any() || string.IsNullOrEmpty(licenseClaims.aud) || string.IsNullOrEmpty(licenseClaims.iss) || licenseClaims.exp == 0)
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = "Incomplete license!"
                };
            }

            if (licenseClaims.iss != validIssuer)
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = $"Invalid license issuer: License must be issued by '{validIssuer}'"
                };
            }

            if (licenseClaims.aud != validAudience)
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = $"Invalid license audience: License audience must be '{validAudience}'"
                };
            }

            // Validate expiration date
            var expiryDate = DateTimeOffset.FromUnixTimeSeconds(licenseClaims.exp).DateTime;
            if (expiryDate < DateTime.UtcNow)
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = $"Invalid license expiry: Your license has expired on '{expiryDate}'"
                };
            }

            if (!licenseClaims.env.Contains(dataverseOrganizationUniqueName.ToLower()))
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = $"Invalid environment: Your license is not intended for usage in '{dataverseOrganizationUniqueName}' but for '{string.Join(", ", licenseClaims.env)}'"
                };
            }

            return new LicenseValidationResult
            {
                IsValid = true,
                License = new License
                {
                    claims = licenseClaims
                }
            };
        }


        public static LicenseValidationResult VerifyLicense(string validIssuer, string validAudience, string publicKeyPem, string licenseKey, IOrganizationService service)
        {
            var key = ImportRsaPublicKey(publicKeyPem);

            if (string.IsNullOrEmpty(licenseKey))
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = "No license key set!"
                };
            }

            // Split the license key into parts
            var parts = licenseKey.Split('.');

            if (parts.Length < 3)
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = "Invalid license format!"
                };
            }

            var encodedHeaders = parts[0];
            var encodedClaims = parts[1];
            var signature = parts[2];

            // Create the data to verify (headers.claims)
            var dataToVerify = Encoding.UTF8.GetBytes($"{encodedHeaders}.{encodedClaims}");

            // Verify the signature
            var isLicenseSignatureValid = VerifySignature(key, dataToVerify, Base64UrlDecode(signature));

            if (!isLicenseSignatureValid)
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = "Invalid license signature: Verification failed!"
                };
            }

            // Base64 decode the claims
            var plainClaims = Base64UrlDecode(encodedClaims);
            var licenseClaims = JsonSerializer.Deserialize<LicenseClaims>(plainClaims);

            var dataverseOrganizationUniqueName = RetrieveEnvironmentInformation(service);

            if (string.IsNullOrEmpty(dataverseOrganizationUniqueName))
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = "Failed to retrieve dataverse organization unique name!"
                };
            }

            var licenseValidationResult = ValidateLicense(validIssuer, validAudience, dataverseOrganizationUniqueName, licenseClaims);

            return licenseValidationResult;
        }
    }
}
